src/Renderer/Label2DRenderer.js

```diff
@@ -0,0 +1,141 @@
+        this.zOrder();
+    }
+
+    renderLabels(object, camera) {
+        // if the object isn't a Label, there are a few options:
+        // - it is visible
+        //      - it has a material but not visible
+        //          - hide its children Label
+        //          - iterate through other children (essentially TileMesh)
+        //      - otherwise iterate through its children
+        // - it not visible, hide all its potential Label children
+        if (!object.isLabel) {
+            if (object.visible) {
+                if (object.material && !object.material.visible) {
+                    object.children.forEach((c) => {
+                        if (c.isLabel) {
+                            c.content.style.display = 'none';
+                        } else {
+                            this.renderLabels(c, camera);
+                        }
+                    });
+                } else {
+                    object.children.forEach(c => this.renderLabels(c, camera));
+                }
+            } else {
+                object.traverse((c) => {
+                    if (c.isLabel) {
+                        c.content.style.display = 'none';
+                    }
+                });
+            }
+
+            return;
+        }
+
+        if (!object.visible) {
+            object.content.style.display = 'none';
+            return;
+        }
+
+        vector.setFromMatrixPosition(object.matrixWorld);
+        vector.applyMatrix4(viewProjectionMatrix);
+
+        if (vector.z < -1 || vector.z > 1) {
+            object.content.style.display = 'none';
+            return;
+        }
+
+        object.x = vector.x;
+        object.y = vector.y;
+
+        // each Label is stored in labels to be sorted in zOrder later
+        labels.push(object);
+
+        if (object.content.parentNode == null) {
+            this.domElement.appendChild(object.content);
+        }
+    }
+
+    /**
+     * The final ordering of labels is done here:
+     * - order the labels by their x/y position on screen
+     * - check the intersection with close neighboors and remove some if they
+     *   are overlapping
+     * - move the remaining labels to their correct positions
+     */
+    zOrder() {
+        // TODO: take into account other values than Y (see mapbox#symbol-z-order)
+        labels.sort((a, b) => -(a.y - b.y));
+
+        let i = 0;
+        while (i < labels.length - 1) {
+            if (labels[i].intersection(labels[i + 1])) {
+                labels.splice(i + 1, 1);
+            } else {
+                i++;
+            }
+        }
+
+        for (let i = 0, l = labels.length; i < l; i++) {
+            label = labels[i];
+            label.content.style.zIndex = l - i;
+            label.content.style.display = '';
+            label.content.style[this.transform] = `${label.anchor} translate(${label.x * this.halfWidth + this.halfWidth}px,${-label.y * this.halfHeight + this.halfHeight}px)`;
```

Comment:
from what you've told me, this transformation is the most consuming...?

Reply:
Yes, but by correctly culling in the process before, we should be able to keep it quite fast (I think like a maximum of 100~ labels updated each time)